### ISSUE TITLE
Gemma 4 VLM: open-ended prompts collapse to <pad> at temperature > 0; duplicate first-token at temperature = 0

### ISSUE THREAD

**Issue by ekryski** [ekryski/mlx-swift-lm#169]
## Summary

The current Gemma 4 VLM implementation (`Libraries/MLXVLM/Models/Gemma4.swift`) misbehaves on open-ended vision prompts in two related ways. The new `--method vision` bench harness from PR #167 surfaced this. We worked around it for the smoke gate by forcing greedy sampling, but it's a real bug that should be debugged before we recommend Gemma 4 VLM for production use.

## Repro

```bash
./scripts/benchmark.sh --model gemma4-e2b --method vision --quick --kv none
# ↑ runs at MLX_BENCH_TEMPERATURE=0 by default — workaround in place

MLX_BENCH_TEMPERATURE=0.6 ./scripts/benchmark.sh --model gemma4-e2b --method vision --quick --kv none
# ↑ reproduces the failure
```

Default fixture: `Tests/Benchmarks/Resources/vlm-test-prompts/test-image1.jpeg` (a golden retriever).
Default prompt: `Describe the image in English.`

## Symptom 1: <pad> collapse at temperature > 0

At the harness default temperature (`0.6`), Gemma 4 E2B 4bit emits exactly one real token (`This`) followed by ~199 `<pad>` tokens (token id 0):

```
Output: This<pad><pad><pad><pad><pad><pad>...<pad>
Validation: Output did not contain expected keyword 'dog'.
Prefill: 439.8 tok/s · Generation: 92.8 tok/s · TTFT: 694ms
```

Same behaviour with the canonical training-time prompt `Describe the image in English` (from `VLMRegistry.gemma4_E2B_it_4bit.defaultPrompt`), so it's not a prompt-phrasing issue.

This shouldn't happen — `<pad>` (token id 0) shouldn't have any meaningful probability in the assistant's response distribution. Most VLMs mask special tokens (pad, BOS, image-token IDs, etc.) out of the sampling pool before applying temperature/top-k/top-p. Either:

1. The logits aren't being masked correctly before sampling, or
2. The chat template is putting the model into a state where `<pad>` is a valid continuation (e.g., the assistant turn isn't properly opened).

## Symptom 2: duplicate first-token at temperature = 0

Greedy decode produces coherent output but starts with a duplicated word:

```
Output: ThisThis is a photograph of a dog, likely a Golden Retriever, standing outdoors in front of a house.
Validation: ✓ (contains "dog")
```

Note the `ThisThis` at the start. The model emits `This` and the very next decode step also picks `This`, producing `ThisThis is a photograph...`. This pattern is consistent with the chat-template prefill leaving a stray token in place — the prefill ends with `...This`, the first decode predicts `This` again, then the model continues coherently.

This points at the same root cause as Symptom 1: the prefill / assistant-turn header is wrong. At temp=0 we get a duplicate (the argmax continuation of the malformed prefix is `This`). At temp>0 the sampling occasionally lands on `<pad>` and then collapses autoregressively.

## Suspected root causes (in order of likelihood)

1. **`Gemma4Processor.prepare` chat-template issue.** Inspect what `tokenizer.applyChatTemplate(messages: messages, ...)` produces for the user-message-with-image case. Compare to the Hugging Face reference processor for `mlx-community/gemma-4-e2b-it-4bit`. Are the `<start_of_turn>` / `<end_of_turn>` / image markers in the right positions? Is the assistant-turn opener (`<start_of_turn>model\n`) appended correctly *after* the user turn?
2. **Special-token masking.** The Gemma 4 attention path doesn't appear to mask non-output tokens (`<pad>`, image-soft-tokens, etc.) before sampling. Confirm by inspecting the logits at decode step 0 — `<pad>` should be `-inf` for the default sampler.
3. **`Gemma4MessageGenerator.generate` ordering.** The `#211` fix in main reordered modalities (images first, then videos, then text). Verify the resulting message structure round-trips correctly through the chat template — particularly whether the trailing `text` content is colliding with the assistant prefix.

## What we did in PR #167

Forced `MLX_BENCH_TEMPERATURE=0` inside `runVisionBenchmark` so the smoke gate produces deterministic + coherent output. The temp override is restored on exit and respects an explicit caller override. This is a workaround — image-description should be sampleable at reasonable temperatures.

## Acceptance criteria

- [x] Gemma 4 E2B 4bit at `MLX_BENCH_TEMPERATURE=0.6` produces coherent multi-sentence output (no `<pad>` collapse).
- [x] Greedy output for the golden-retriever fixture starts cleanly (no `ThisThis` duplicate).
- [ ] Special-token masking is in place: `<pad>` and image-soft-token IDs have `-inf` logits at decode step 0.
- [ ] Remove the `MLX_BENCH_TEMPERATURE=0` workaround in `runVisionBenchmark` once the underlying bug is fixed; bench reverts to the harness default (currently 0.6).
- [ ] Add a unit test on the `Gemma4Processor.prepare` output token sequence (compare structure / tag positions to a recorded fixture).

## References

- Workaround commit: `008abc5` (PR #167 / `ek/merge-main-into-alpha`).
- Related: #168 (LLM/VLM common-code consolidation — Gemma 4 VLM perf reconciliation).
- Tokenizer / chat template plumbing: `Libraries/MLXVLM/Models/Gemma4.swift` lines around `Gemma4Processor.prepare` and `Gemma4MessageGenerator.generate`.

**Comment by ekryski** [comment]
Linked to #168 (LLM/VLM common-code consolidation). Fix this one first or alongside the consolidation so the perf-reconciliation work isn't comparing against broken sampling.

**Comment by ekryski** [comment]
## Resolution — fixed in [4f6c9b7](https://github.com/ekryski/mlx-swift-lm/commit/4f6c9b7)

Both symptoms (`<pad>` flood at `temp>0` and `ThisThis` duplicate at `temp=0`) had the same root cause: **`Gemma4VLMModel.prepare(_:cache:windowSize:)` returned `.logits(result)` without imposing an eval barrier on the cache state**. The iterator handles `.logits` by sampling the first token from `result.logits` and immediately starting decode — but the first decode forward reads K/V from the cache before the prefill's writes commit, producing garbage logits at the boundary.

This matches the LLM-side prefill-sync barrier already in `MLXLLM/Models/Gemma4.swift:prepare` (which uses `asyncEval(cacheArrays)` after each chunk of its chunked prefill). The VLM does single-shot prefill so we drop one hard `eval()` at the end.

```swift
var cacheArrays: [MLXArray] = []
for c in convertedCache {
    cacheArrays.append(contentsOf: c.innerState())
}
eval(cacheArrays + [result.logits])

return .logits(result)
```

**Tested both async and sync — async was insufficient.** With `asyncEval(cacheArrays)`, the writes were still pending when the iterator's first decode read happened. The hard `eval()` barrier is required to force the metal command-buffer flush before returning.

### Verification

`mlx-community/gemma-4-e2b-it-4bit` on alpha @ `4f6c9b7`:

| Run | Output | Decode tok/s |
|---|---|---|
| `--method vision --temp=0` (golden retriever) | `"This is a photograph of a dog, likely a Golden Retriever, standing outdoors in front of a house."` | — |
| `--method vision --temp=0.6` | `"This is a photograph of a beautiful, long-haired dog, likely a Golden Retriever, standing outdoors..."` | 90.7 |
| `--method simple` (greedy text) | `"Hello! I am a helpful assistant..."` | 94.9 |
| `--method summarization 1024` (text) | `"The provided text is an excerpt from **The Great Gatsby**..."` | 95.9 |

No prefill/decode tok/s regression. Bench harness's `priorTemp` workaround stays in place for bit-deterministic across-day report rows, with the comment updated to reflect that the underlying pathology is fixed.

### Follow-up

WS-C #7 (the Gemma 4 LLM↔VLM consolidation PR) will lift this barrier into the shared `MLXLMCommon/Models/Gemma4.swift` namespace once both target's language-model classes converge.
